{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fa721720",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "from tensorflow.keras import datasets\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "53757562",
   "metadata": {},
   "outputs": [],
   "source": [
    "(x_train_raw, y_train_raw), (x_test_raw, y_test_raw) = datasets.mnist.load_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1b626670",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "(60000, 28, 28)\n",
      "(10000, 28, 28)\n"
     ]
    }
   ],
   "source": [
    "print(y_train_raw[0])\n",
    "print(x_train_raw.shape)\n",
    "print(x_test_raw.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "94957609",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n"
     ]
    }
   ],
   "source": [
    "num_classes = 15\n",
    "y_train = tf.keras.utils.to_categorical(y_train_raw, num_classes)\n",
    "y_test = tf.keras.utils.to_categorical(y_test_raw, num_classes)\n",
    "\n",
    "print(y_train[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6a859dde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAd4AAAGFCAYAAABT15L3AAAAPHRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMHJjMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/RjVi6AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAf1ElEQVR4nO3daXhUVbr28V2phJCEMcFEwDCGGAQEBBQQxQEU3xfBAUTUV8SpcQClcThyelbP0da2BQRnwKEbPc62ttKiyLFlEBRQGcIcRsM8JoGkqs4Hr7fCvU+nSEjqqen/+7Rv965dq/uq8GTlqbW2JxAIBBwAAGAiKdIDAAAgkVB4AQAwROEFAMAQhRcAAEMUXgAADFF4AQAwROEFAMAQhRcAAEPJ1b1wYNLwcI4DYfCZ/61IDyEq8NmNPXx2f8ZnN/ZU57PLjBcAAEMUXgAADFF4AQAwROEFAMAQhRcAAEMUXgAADFF4AQAwROEFAMAQhRcAAEMUXgAADFF4AQAwROEFAMAQhRcAAEMUXgAADFF4AQAwVO3n8QKIbhUX9ZC8486jkpf3eUVy1wWjJLeYWk+yd+53dTg6AP8fM14AAAxReAEAMJQQf2r2JOv/TO8pzar92sL72kj2pfslt26/U3L6nR7JPz2lf777ruebknf7jkg+560JweO8Xy6s9jiRePz9u0uePP0ZyXkp+rnXT67jLO0zQ3JhT5/k+9v0rt0AgQg5MuwcyY//8VnJD19zo+TAkh/DPqbjMeMFAMAQhRcAAEMUXgAADMVEj9fbsYPkQGqK5O39m0gu7a1908zGmr/qqn3W2vikpKHkx58ZJHlRl79K3lheKvmx4oGSW3wVqLOxIf6UX9IzePzAtNfkXH6Kfp/A7+rqbigvl3zAnyq5u0bn6GW9JKfN/UHvX1Z24gEjokqHnl15nOWVc5nTF1gPx8zOnjqnfHjT5REayb/GjBcAAEMUXgAADFF4AQAwFJU9Xt8FZ0l+auZUye5elqXygK51/M2UmyQnH9EebZ+37pbccFuF5NTd2vNNX7KoliNELPM2aiT5yPkFksf/ufI7AxemHXa9OvTv0TP39ZX8+bQ+kr/+3WTJn730nOQzXtfPcrsH47dHGC+2n1/5mUhvv19PTrcdS1glaf860Er/Xb04e7Xkzz36s2CNGS8AAIYovAAAGKLwAgBgKCp7vKmF2yV/W5YrOT+luM7ea8IO3Y92w2Hdx3lm+7clH/BrDzdn8vxavT+rdnG8ra+2lLy419Qqrqy5P2QvlvxpA+1zjd50ieRX2syR3OiMPXU2Ftj4/eC3gsePr7okxJWxzdu+teTV/bWB3e2bGyS3WKxr0q0x4wUAwBCFFwAAQxReAAAMRWWPt2LHT5KnPD5c8qODdO9l7/cNJC+/c0rI+z+y+8zg8boB6XLOt3+H5Ov63Cl50zi9V1tnecj3AkKpuKiH5Fnd9Jm6SU7Va9ZHF10secmcjpJ/uEXvNbe0vuTsJbrWcd0+XTOc8h9zdSz6qGnEgBRPxYkvigPJL5WEPF+6vlHI89aY8QIAYIjCCwCAIQovAACGorLH65Y5Q/eEPeVvWZJ9e/ZK7tT5Zskrztc1XR++0D94nL0/9DpczwLt4bZle1rUgr9/d8mTp2sfNi9FfyTdz9QdsvrK4LF3mH7Xocn/1VXhZ7ymeyvnT90iOWnLUslNv9Kxlj+q+5K/c6b+HN18oX7hwTv3OweR5e/XTfJ59f8ZmYEYa5MReo157hxfyPPWmPECAGCIwgsAgCEKLwAAhmKix+vm2x367/nlB0M/r7fT9SuDx7ue1ec4Ov7o6gUgtnl6dJK8+5e6dtb9bOlvj+rrvzh8huQ9b1TuW561T79w0Pj1hZpdY6ntis4cb6qO5V5dO5mty34RAUWD0yRne9OruDK2JbdpJXlY5ochr0/buE9ypP+VZ8YLAIAhCi8AAIYovAAAGIrJHu+JdHxwjeTRXXRP2xmtPw8e9x9+l5xr+Kb2yYCaSErXnlrFHw9KXljwruSNFcck/3LiBMlNv9osOTtjZ/A40n2qs5sXSd4UmWHgOMl5h6o8V7a6id1AwmzL0xmSz03V9e4vHzxNX7Bffw4jjRkvAACGKLwAABii8AIAYCgue7y+/Qck77lDn1O6+cPKtZT/9sircu6ha66UHFiqqyFzH3Vt1hzQ/XGR2Er767rd2QXTQl5/6z3jJTd8X79jkBhPU4WF7CX+E18UId5muv9+8dX5kjOv2Sp5Xv7Lrjvos6afnXqF5Ozi0HvyW2PGCwCAIQovAACG4vJPzW7+5askX/v7+4PHf/ntk3JuWW/907PTW2OnDH3UWocXd0iu2LDp5AaJuHDmw8skJ7l+tx1dpEvb0t7/JtxDOmkpHt1OtdzVVfF6aLPEktJM/SxmVHFdVfzn6SMtA16P5C0DdEvRYy3KJSfVq1wA94/zpsi5FL2V85NP7/XrDdoC3OvXP5unJ+niupxFuqwq2j6pzHgBADBE4QUAwBCFFwAAQwnR43XLnF65JOjuQt0ystFj+rX1We1mS15x4zOSC3JvlXz67/V3Gd/aDSc9TkS//f+vj+Rf5eh3BvyO67F//9DH/LVyomuZw/HKA9o38zvaV/t0lf5v6eB8F/YxIbSjZSmS/cd1N2dM/LOc+/DubjW694NZL0lOcrQxWxrQ7U+3+/Tz88yuC4LHA+bcK+eaLNWfk+b/KJbsKdJ/l3et0scf5ni1nxxY/IMTzZjxAgBgiMILAIAhCi8AAIYSssd7PM/XyySXDMuW3GvEWMmLHpwkefWF2ve4vs0lkg/0q+UAEdUqtNXkNE7SXtWCMl2P2O7V7fr6sIyqetyPMFz9ZGfXFd9Kun7DZZIL7tkoOdKPKYTj5N2wVHKn/6zcdyC317Za3XvuTt3Gcdcn+ui9rBXaZ6336WLXHSrP5ztLQr6X+7O07cG+knul6ta9bxxuGfJ+0YYZLwAAhii8AAAYovACAGAo4Xu8br7inZJzJmsue0C7cuke7em92OYjyYOvvFevf29RLUeIWLLH10ByJPfydvd0Cx/rInn1UF2j/kmJPhJz+9Q8yQ336SMMEX3aPrTgxBedpObO5rDd2y39/F0hz/9q7tWS853o3QPdcZjxAgBgisILAIAhCi8AAIYSvsfr79dN8vrh9SV37rZJsrun6zZlrz6zMv2D0OvVEN/u+3q45HzX2thw8vfXz+LOX5ZKXtVTe7oX/zBCcsYg3We8oUNPF9Gp9QfR9sTd0JjxAgBgiMILAIAhCi8AAIYSosfr6al70K4ZV9mnffHcV+Tc+fX1mZIncjSg+5Mu3NtWL/DvqNH9EGP0kaROkut32Un9Zkme6uh+t3Wp6A/6bOB3bnxKcn6Kfj/hrG9GSW5x5crwDAyAYMYLAIAhCi8AAIYovAAAGIqLHm9y29aS149uIfl3I96QfHWD3Sf9XhOLe0qeN6m35KavhG9vVEQh1/JBv+OX3D9tj+R7Z/aQ3H6GXp/y0yHJxf1PkZw5YmvweGyrz+XcZem6RvjDIzmSb/xhkORmz2c4QCzyenTOuC8/RfKpn1iOpuaY8QIAYIjCCwCAIQovAACGYqLHm9ymleQDPZpLHvGHTyWPafLuSb/XhB3as10wTXu6mTP1OY9N/fR0UbX6Hv0RWzXwOcn/PE/3Bl979FTJoxtvqvZ73bP9PMmfzu8mucM97LWM+OAL6HcjYm0KGWPDBQAgtlF4AQAwROEFAMBQVPR4k5trX2vvdF1feEfbeZJHNiyu1fvdva1f8Pi7Z7vJuWZv/yg58xA9XFQt58udkh/8he6X/PipoT8/7r3B+9XfFPL6pUcrf1ceOe92OZc/WtfxduD5uUgQJb1KIj2EGmHGCwCAIQovAACGKLwAABgy6/Eeu7RyPeyx8Xvl3MS8v0u+JO1Ird6r2Fcq+fwPJ0gu+NXq4HHmfu3BuVaHASH51qyXvHZ4G8lnjB0reeU1U2p0/4K/3yn59GmVvaz8pd+6LwcSgnuv5lgT26MHACDGUHgBADBk9qfmTVdU1vg1Xd6q0Wun7m8vedK8SyR7fB7JBY9slNyheJFkX43eHai+ig2bJOeN1zxkfK8a3S/fWSw5UMV1QDw7Okcfj+nrFttNQWa8AAAYovACAGCIwgsAgCFPIBCoVttoYNLwcI8Fdewzf8166fGKz27s4bP7Mz67sac6n11mvAAAGKLwAgBgiMILAIAhCi8AAIYovAAAGKLwAgBgiMILAIAhCi8AAIYovAAAGKLwAgBgiMILAIChau/VDAAAao8ZLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhpKre+HApOHhHAfC4DP/W5EeQlTgsxt7+Oz+jM9u7KnOZ5cZLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYovACAGCIwgsAgCEKLwAAhii8AAAYSo70AGLd+if6SF513TOSUzxeyeffebvktPe/Cc/AACBGeLMyJXsaN5K8+eoWksuaBSTn/X65ZH9JSR2Oru4x4wUAwBCFFwAAQxReAAAM0eOtoZ/G95X85Yg/Si4P1At9g0Do0wAQj5I6FwSP1z6UJudu7jJf8oSs2TW6d8ecMZI73PRtDUdnixkvAACGKLwAABii8AIAYIgebw0dzvVLzkw6QU8XqIVjl/aUXHR95efvjrPmybl7m64Jea8uL42VnL5Dv3Cwv+9Rya3/or+X15u9JPRgkdA8vbpIXjde9zD4sl/lHgeneFPlXJJrDvhxSVPJG45mS76raaHk185/UfLDvUZJDiz+oaphRwQzXgAADFF4AQAwROEFAMAQPd4TODz8HMnvXDnJdYVH0nP7CyTPuUZ7dBlFKyRrxxiJbtcY3ft7ygNTJfdM9QWP3X2xUZsGSO7eeLPk5be6P7vKfb++mSMlZ9ZsaSXijPeUUySvmdRS8t/6TpPcLiXFdYdUpyozDuZKfv/qfpL9qXqvuz7SHu/xPxeO4zilObpOuH6V7xwZzHgBADBE4QUAwBCFFwAAQ/R4XcoGny35t/85XXJ+ivZ03V55cZDkU1fOr+JKJCJPiq77LhvQVfI7Dz0huUWy9sVuKRoYPC568nQ5l/HxMslz01tJnvdevr5Xhw9DjvXgsizJmVVch8Sw7YYOklf0d39nwN3Trdrr7p7uFboHvq9Q16R7uneq9r1jATNeAAAMUXgBADBE4QUAwBA9XpcdN5RJvjCtzHWF7j/qXjt56iR6uqjajrt1Xfc397n7ZNrTHb7ucskVV5cHj9N3L5Jz7kc9b7+9h+RFHUKv4/2kpKHkvOe36HuHfDXiXcshm2p0/duHT5X81JqLg8c5D+in1Ve4NuS99nVpVKP3jnbMeAEAMEThBQDAEIUXAABDCd/jTT5N9xtdcd4MyeUB3QN0VblEZ/NTujYyw9G+GxLb2im613fhVVMku/fq7vjZGMkF922S7Nu9p9rvPeaOD6p9reM4ziOP6jNMm25ZUKPXI87dpt8/OOMufb5z7mf6b2XGip8kNyuqXJurV55YSU7o/RNiDTNeAAAMUXgBADBE4QUAwFBC9ni9nSr3uO351x9r9NoR746T3P6dhXUyJsSH9X/qLbnwKn2e7gG/rgsfvvo6yaeP1T1qfYcOVfleSRkZkvcMO1Py0Aa673OSo88oLXjrLsl5M+npomq+dRsl543fWMWVP6vLdd/lvar+OYhFzHgBADBE4QUAwBCFFwAAQwnZ4y0aUvmc0bezlrrO6l7M163XvXLzH1svuabr0RBfvDnZkl+5cppkv2ulrrunW29gkev60JK6nRE87jx9lZx7JGey62pdd3nusmsln/47fT2fZYTT5t9UPnO3It21s7h7ma7r9FUdQn//4O6tF0hO+/S7ULeLOGa8AAAYovACAGAoIf7UvHd0H8nvjTl+mUWKnBuzpb/k8lH65zrfrs11OjbENk99/Xz0TA39B9u0cfX09a1zJa8dc5rkSwbon8zGZ78QPG6VrMuD3H+m9gX0D2yeN5vp+f2hH8UGhOJtpI/qKzu7g+SUh4olf1+g26XKtR5t8bm36nWbW5oueevtrSQHKrSNEm2Y8QIAYIjCCwCAIQovAACG4rLHe/yWkI7jOPMfecZ1Rf0qX7tgaxvJuZtqtqUkEkug7KjkRUf1OwPnpOpzJD+Y84Zk93KjE5lTWtmnXVuuPdwL0w5LXnJM+8lNXmVLSFSfJ1W/v3CsfxfJ46e9JvnCtM8lF/v0Z2NuadPg8W/WDJVzszrNlNwiWd/brX6S/lxtuKaJ5HaF+m+8v0y3ao00ZrwAABii8AIAYIjCCwCAobjs8a6ZqGu8TrQm7HitHtMcbVuNIbr4indK/u0dt0p+8jndQvJMbbs6rx/UdbyPzBsiOX+m9qaSiw8Ej7Nn7ZVzF+Z+IXnUXB1LvrPEAaqSVF/7ontGdJf81X+4tyRVnWaNlXzaXP13N/XjxcHjrOb6fYRZs3tInpAV+rs17u9OfH+Tjq3PFn18a86ryyX7S0pC3j/cmPECAGCIwgsAgCEKLwAAhuKix+vvr72IR3q+X+3XDvxRH5XWYAnrdnHy6s3WPurEtmfX6PX5zjchzx8aWnm/j1t9IOfKA/p7dNomV0MZOI57ne7qp87UPDR0T3do4RWS85/YINn9/Yfk3Mp9yLt+qHve35+1UvIB/zHJ57wzQXLzAr33513elLzg1zr2ESMHS949Wdck19+jPePjeb/8rspzJ4sZLwAAhii8AAAYovACAGAoLnq8j858QXLnlNCrb+/bcX7wuPHIfXKu+it+AXsVaZW/K7vXp7v3fW47U/toFeEbFmKAJ1n/uS98uqvk1UOmSt5aoXstD3n+Acltpq+XXOHq6ZYP0LW5nR9fGjz+bfa3cm7GwdaSX/v3yyXnvbtQsrdZluQLBuoa4iMjDkh+r/uLkk+bXPVe0B8d0Xu/kN+uymtPFjNeAAAMUXgBADBE4QUAwFBc9Hi719PfH060N/OCGWcFj7P3zQ/LmIBwaPjGcb2uP0VuHIg9W+7XNeWrh0ySvN3V0x3+2P2S27yv63T3XtRWcuCGhpLf7qz3P8Vb2Vft9Ib2ZPNf2C05vXCRE4pv9x7JjWa5s14/7E7tT+cMK6r65hOauP7DipBjORnMeAEAMEThBQDAEIUXAABDMdnj3fJ2Z8kpnmU1en3zLyv7CazbRSw5dG3v49K3VV4HuD1727SQ5+t7NF8+5r8ltxynex6MavS3E7yjrpXt9NfKZ+TmPbRYzvkqwrvKPHuafpcnEPL/im1hHYvjMOMFAMAUhRcAAEMUXgAADMVEj9f9vN2nu70u2b1u94C/THKvT+6VXFCkz34EYsWBdvyujJPz34cLJJ+T+oPkTK/2ZCc2WxbyfoNXXyV584LTJLd7W/dLzltR+Z2EQJh7utGOn2IAAAxReAEAMEThBQDAUEz0eMsy60nuV/+I6wqvpNklrSTn365rxvSppUDsaDmvJHiccrd+7stDP4YaCW7+hS0kn3P9RZIPdD0mOXlXiuT853R9a/JP+vzdNmVbJPPvbNWY8QIAYIjCCwCAIQovAACGYqLHC+Bnnq+XBY9nHsyWcyMbag+upFNzyfW2bA3buBD9fHv2Ss6ZrPsX55zg9Ym98rZuMeMFAMAQhRcAAEMx8afmRst+kjx2q34N/rnceZbDAaLCn58fJnnkfZMkN//1Osl79p+pN1j4fVjGBSA0ZrwAABii8AIAYIjCCwCAoZjo8VZsLJK8tbeeH+z0MBwNEB1avlYoecQVgyW/mfeR5P6/GSk587rGkn379TFuAMKDGS8AAIYovAAAGKLwAgBgKCZ6vAD+N9/uPZKPXZ0lueOffiF51YDnJQ8puEVvyLpewAQzXgAADFF4AQAwROEFAMAQPV4gTrh7vh1GaR7i9HK9gp4uEAnMeAEAMEThBQDAEIUXAABDnkAgEIj0IAAASBTMeAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAwlV/fCgUnDwzkOhMFn/rciPYSowGc39vDZ/Rmf3dhTnc8uM14AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAEIUXAABDFF4AAAxReAEAMEThBQDAULWfxxvL1szoIXnjpS8Hj5/a207Ozbmmp2TfyjXhGxgAIOEw4wUAwBCFFwAAQ3H5p2Zvp9Mlf3DhVMnlgZTg8V1NC+Xc22deIrnhyjoeHBCCp0cnyf56+iO67YKM4PGKsdPkXHnAV6djufjHYZIzhu7QsZWV1en7Ib54UlMll1zWNXh85r8vl3Nrex01GVO0YMYLAIAhCi8AAIYovAAAGIrLHq+z7SeJ49ZcK/mzTu9YjgYICvTpKnntTfUk//miWZJTPBWSB6QdCh6XB/T3Zr/jr4shBn3W+b8kd3vtZslt79gu2bd7T52+P2Kb95RmkudOfS54/FWZlp4n2l4uuWJjUfgGFgWY8QIAYIjCCwCAIQovAACG4rLH69t/QHLR1g56gS6VBMwEHtkreXXBuxEaSc0t6ztd8qXn3Ck59WN6vKie8+rrdxcebZUpOYkeLwAAqCsUXgAADFF4AQAwFJc9Xm9OtuTzOvJoP0SHbV/m6n8oCH39gjLd7/bmv99WGTyuiwOh79X7LP05mNHmH6FfAISJ15PYc77E/l8PAIAxCi8AAIYovAAAGIrLHq/TMEPi/8lcXO2X7uyhjbMm3+dL9q2kX4yT1+qxJZKv/K+RIa/3HCuX3GHjopN+7/3NsiTPWdhQ8vH7QP8rF/0wQnKjuSsk1+1O0YhnvoB+WsrTtRTpNxviDzNeAAAMUXgBADBE4QUAwFBc9nh96zZK/tXftDd19cipVb52xXWTJXc/cI/kXHq8qIVA+THJvsJ1Zu9dfJV+X6FLvQ9cV4TurG3frvvpNijZUBfDApydPVIk534SoYEYYcYLAIAhCi8AAIYovAAAGIrLHq9b+/sW6n8IvXQSiAu77ugjueCG1ZJzvDVbLdnxAf3uhO/khoUEESjXNehrysuCx/kp9eVcaVv97kO8Y8YLAIAhCi8AAIYovAAAGEqIHq9biscbPC4/wTNMgWi18+6+kkfd8XfJNzR6UnLDpHo1uv/Du86SHDiaWH041I6veKfkcesr91P4tMC9hjyxMOMFAMAQhRcAAEMUXgAADCVkj7c8ULkC0c9TRGHI2+l0yWtGN5Xcv9+P1b7XR7lTJP/vz3Lonu668grJI56dILnVe8V6/0Prqz02AFVjxgsAgCEKLwAAhhLyT82AlcC53STfNOM9yUMzdtfi7rX7vXncOn1cZsvH50tmS0hYaZBZEukhmGLGCwCAIQovAACGKLwAABiixwsY8jq6R2lSLX73PX7rU8ep+fann3bUfvN5198lufFfXI/TBMLknbNelDzWOTdCI7HBjBcAAEMUXgAADFF4AQAwlJA93po8FrBR352hLwBC8Hy9TPLLVwyS/G83ZUluNVsfvect1W0da2LtLSmSVw969qTvBdTWln/mVoaCyI0jGjDjBQDAEIUXAABDFF4AAAwlZI+3Jo8FnNd1luQhvW/RCxZ+X2fjQvzzrVwjud0D4XuvjmtP0f8w6F9fB1hosKXqL9Q09Og57xn5kt0/N7GOGS8AAIYovAAAGKLwAgBgKCF7vAVf3Bo8XnnRCzV67Zrb60nOZztbRKniq/IiPQQgKCnEknSvxyPZn5ZSxZXxgRkvAACGKLwAABii8AIAYCghe7ypa9Iqw0WRGwdinyc1VfL+4d0lN/1ghWT/oUNhG8uOCX0lfzDuj64rUh0gUprOXBA8fu6B1nJuTOMiyWvH63dp8m4I37gigRkvAACGKLwAABii8AIAYCghe7y5D88PHs+6vqWcu77hjpCv3TjoJcmXdR0p2b98VS1Hh2hWdvnZkhvft1nyvLwpkq9crJ8Pp7B2Pd7k5qcGj7cNayfn3hz7pOQWyaF7usW+o5JTSk/wcGqgjjy58FLJgy5+WnL+L3Rv5tA76sceZrwAABii8AIAYIjCCwCAoYTs8R5v5mZd+ziy01shry+nDZbQLn10nuQJWT+GvH71xEb6Hw6fU6v3v7Zv5VrI97M/lnN+J/T+tqM2aV9t3YzTJWe9u8ABIsHnuPZqLi2L0EhsMOMFAMAQhRcAAEMUXgAADCV8j/fozFP1PzwRmXEgPq0a8HwY766/Ny8o03W7ty26UXLebWslZx2hp4vo0D45TfKe0bpePuvl+PqsMuMFAMAQhRcAAEMUXgAADCV8j7fpsr2Sp+7TtY13NS20HA6i3BfjzpX86p3ai1p+7vQ6fb/XD+ZK3lHeJHg8/TsdS96LPsntvl4mOd72u0XsmtFff072+UslN/v+sOR42z6BGS8AAIYovAAAGEr4PzX7Vurjp2Z31i3+Zju9TnAHHgOYSLxffie57TfpknuMu0fyK794WnLnero13kU/jJB84Etd3tb6zW2SKzYWBY87ON+ecLxANLp/1TDJw1ovlZx0RB9ZqU2U2MeMFwAAQxReAAAMUXgBADCU8D1eoDb8JSWSWz42X/LEx3S5kVsDZ0PIXFGLsQHRKnOwfrfmCyfDdcUaJ54x4wUAwBCFFwAAQxReAAAMUXgBADBE4QUAwBCFFwAAQxReAAAMUXgBADBE4QUAwBCFFwAAQxReAAAMeQKBQCDSgwAAIFEw4wUAwBCFFwAAQxReAAAMUXgBADBE4QUAwBCFFwAAQxReAAAMUXgBADBE4QUAwND/AMpcvRyG0jCyAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 9 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "for i in range(9):\n",
    "    plt.subplot(3,3,i+1)\n",
    "    plt.imshow(x_train_raw[i])\n",
    "    plt.axis('off')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1862a07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train = x_train_raw.reshape(60000, 784)\n",
    "x_test = x_test_raw.reshape(10000, 784)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d9a23f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train = x_train.astype('float32')/255\n",
    "x_test = x_test.astype('float32')/255"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b7c5d3e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential_1\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " dense_5 (Dense)             (None, 512)               401920    \n",
      "                                                                 \n",
      " dense_6 (Dense)             (None, 256)               131328    \n",
      "                                                                 \n",
      " dense_7 (Dense)             (None, 124)               31868     \n",
      "                                                                 \n",
      " dense_8 (Dense)             (None, 64)                8000      \n",
      "                                                                 \n",
      " dense_9 (Dense)             (None, 10)                650       \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 573,766\n",
      "Trainable params: 573,766\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "model = tf.keras.Sequential([\n",
    "    tf.keras.layers.Dense(512, activation='relu', input_dim=784),\n",
    "    tf.keras.layers.Dense(256, activation='relu'),\n",
    "    tf.keras.layers.Dense(124, activation='relu'),\n",
    "    tf.keras.layers.Dense(64, activation='relu'),\n",
    "    tf.keras.layers.Dense(10, activation='softmax')\n",
    "])\n",
    "\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0ecbae1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8e90ef7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "in user code:\n\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1160, in train_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1146, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1135, in run_step  **\n        outputs = model.train_step(data)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 994, in train_step\n        loss = self.compute_loss(x, y, y_pred, sample_weight)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1052, in compute_loss\n        return self.compiled_loss(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\compile_utils.py\", line 265, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 152, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 272, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 1990, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 5529, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (32, 15) and (32, 10) are incompatible\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn [23], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepochs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m10\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\utils\\traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     67\u001b[0m     filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[0;32m     68\u001b[0m     \u001b[38;5;66;03m# To get the full stack trace, call:\u001b[39;00m\n\u001b[0;32m     69\u001b[0m     \u001b[38;5;66;03m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m     72\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[1;32m~\\AppData\\Local\\Temp\\__autograph_generated_file9_veblae.py:15\u001b[0m, in \u001b[0;36mouter_factory.<locals>.inner_factory.<locals>.tf__train_function\u001b[1;34m(iterator)\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     14\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m---> 15\u001b[0m     retval_ \u001b[38;5;241m=\u001b[39m ag__\u001b[38;5;241m.\u001b[39mconverted_call(ag__\u001b[38;5;241m.\u001b[39mld(step_function), (ag__\u001b[38;5;241m.\u001b[39mld(\u001b[38;5;28mself\u001b[39m), ag__\u001b[38;5;241m.\u001b[39mld(iterator)), \u001b[38;5;28;01mNone\u001b[39;00m, fscope)\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[0;32m     17\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[1;31mValueError\u001b[0m: in user code:\n\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1160, in train_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1146, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1135, in run_step  **\n        outputs = model.train_step(data)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 994, in train_step\n        loss = self.compute_loss(x, y, y_pred, sample_weight)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1052, in compute_loss\n        return self.compiled_loss(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\compile_utils.py\", line 265, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 152, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 272, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 1990, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 5529, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (32, 15) and (32, 10) are incompatible\n"
     ]
    }
   ],
   "source": [
    "model.fit(x_train, y_train, epochs=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9c517ca5",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "in user code:\n\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1727, in test_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1713, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1701, in run_step  **\n        outputs = model.test_step(data)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1667, in test_step\n        self.compute_loss(x, y, y_pred, sample_weight)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1052, in compute_loss\n        return self.compiled_loss(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\compile_utils.py\", line 265, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 152, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 272, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 1990, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 5529, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (None, 15) and (None, 10) are incompatible\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn [24], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mevaluate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx_test\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_test\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\utils\\traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     67\u001b[0m     filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[0;32m     68\u001b[0m     \u001b[38;5;66;03m# To get the full stack trace, call:\u001b[39;00m\n\u001b[0;32m     69\u001b[0m     \u001b[38;5;66;03m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m     72\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[1;32m~\\AppData\\Local\\Temp\\__autograph_generated_filecgjxyf5z.py:15\u001b[0m, in \u001b[0;36mouter_factory.<locals>.inner_factory.<locals>.tf__test_function\u001b[1;34m(iterator)\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     14\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m---> 15\u001b[0m     retval_ \u001b[38;5;241m=\u001b[39m ag__\u001b[38;5;241m.\u001b[39mconverted_call(ag__\u001b[38;5;241m.\u001b[39mld(step_function), (ag__\u001b[38;5;241m.\u001b[39mld(\u001b[38;5;28mself\u001b[39m), ag__\u001b[38;5;241m.\u001b[39mld(iterator)), \u001b[38;5;28;01mNone\u001b[39;00m, fscope)\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[0;32m     17\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[1;31mValueError\u001b[0m: in user code:\n\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1727, in test_function  *\n        return step_function(self, iterator)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1713, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1701, in run_step  **\n        outputs = model.test_step(data)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1667, in test_step\n        self.compute_loss(x, y, y_pred, sample_weight)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\training.py\", line 1052, in compute_loss\n        return self.compiled_loss(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\engine\\compile_utils.py\", line 265, in __call__\n        loss_value = loss_obj(y_t, y_p, sample_weight=sw)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 152, in __call__\n        losses = call_fn(y_true, y_pred)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 272, in call  **\n        return ag_fn(y_true, y_pred, **self._fn_kwargs)\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\losses.py\", line 1990, in categorical_crossentropy\n        return backend.categorical_crossentropy(\n    File \"C:\\Users\\ASUS\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\keras\\backend.py\", line 5529, in categorical_crossentropy\n        target.shape.assert_is_compatible_with(output.shape)\n\n    ValueError: Shapes (None, 15) and (None, 10) are incompatible\n"
     ]
    }
   ],
   "source": [
    "model.evaluate(x_test, y_test)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
